### PR TITLE
add `isNamedTuple`; make $(1, 2) be (1, 2) instead of (Field0: 1, Field1: 2) which leaked implementation detail 

### PR DESCRIPTION
**Pull request by timotheecour**
* fixes this: https://github.com/nim-lang/Nim/pull/8670#discussion_r211156536 /cc @alehander42 @Vindaar @mratsim 

* Note: `isNamedTuple` is useful in other places, eg https://github.com/nim-lang/Nim/pull/10010#issuecomment-448013600

* ~~depends on https://github.com/nim-lang/Nim/pull/10080 (for the diff) ; so please don't merge this PR before it~~ 
EDIT: ~~removed dependency on this by modifying the test~~
EDIT: now that #10218 was merged, modifying that test wasn't needed anymore (as far as static[T] issues are concerned, that is; t still needed patching to reflect the new behavior introduced in this PR)

### CODE REVIEW

File: lib/system.nim [hunk 1]
```diff
@@ -2651,19 +2651,40 @@ proc compiles*(x: untyped): bool {.magic: "Compiles", noSideEffect, compileTime.
   ##     echo "'+' for integers is available"
   discard
 
+proc isNamedTuple*(T: type): bool =
+  ## return true for named tuples, false for any other type.
+  if T isnot tuple: result = false
```
Comment (Vindaar) [inline lib/system.nim]: This could be a `when T isnot...` no?

File: compiler/types.nim [hunk 2]
```diff
@@ -1172,7 +1172,7 @@ proc typeAllowedAux(marker: var IntSet, typ: PType, kind: TSymKind,
   result = nil
   if typ == nil: return
   if containsOrIncl(marker, typ.id): return
-  var t = skipTypes(typ, abstractInst-{tyTypeDesc})
+  var t = skipTypes(typ, abstractInst+{tyStatic}-{tyTypeDesc})
```
Comment (Araq) [inline compiler/types.nim]: This change is unrelated to your PR.

File: lib/system.nim [hunk 3]
```diff
@@ -2651,19 +2651,40 @@ proc compiles*(x: untyped): bool {.magic: "Compiles", noSideEffect, compileTime.
...
   discard
 
+proc isNamedTuple*(T: type): bool =
+  ## return true for named tuples, false for any other type.
+  when T isnot tuple: result = false
+  else:
+    var t: T
+    for name, _ in t.fieldPairs:
+      when name == "Field0":
+        return compiles(t.Field0)
```
Comment (Araq) [inline lib/system.nim]: Huh? When would that not ``compiles``?

File: lib/system.nim [hunk 4]
```diff
@@ -2651,19 +2651,40 @@ proc compiles*(x: untyped): bool {.magic: "Compiles", noSideEffect, compileTime.
   ##     echo "'+' for integers is available"
   discard
 
+proc isNamedTuple*(T: type): bool =
```
Comment (Araq) [inline lib/system.nim]: Use ``t: typedesc`` here.

File: lib/system.nim [hunk 5]
```diff
@@ -2651,19 +2651,40 @@ proc compiles*(x: untyped): bool {.magic: "Compiles", noSideEffect, compileTime.
...
   result = "("
   var firstElement = true
+  const isNamed = T is object or isNamedTuple(T)
+  when not isNamed:
+    var count = 0
   for name, value in fieldPairs(x):
     if not firstElement: result.add(", ")
-    result.add(name)
-    result.add(": ")
+    when isNamed:
```
Comment (Araq) [inline lib/system.nim]: Move this check out of the ``for`` loop.

File: lib/system.nim [hunk 6]
```diff
@@ -2651,19 +2651,40 @@ proc compiles*(x: untyped): bool {.magic: "Compiles", noSideEffect, compileTime.
...
   ## .. code-block:: nim
-  ##   $(23, 45) == "(Field0: 23, Field1: 45)"
+  ##   $(23, 45) == "(23, 45)"
+  ##   $(a: 23, b: 45) == "(a: 23, b: 45)"
   ##   $() == "()"
   result = "("
   var firstElement = true
+  const isNamed = T is object or isNamedTuple(T)
+  when not isNamed:
+    var count = 0
```
Comment (Araq) [inline lib/system.nim]: When you do it right, there is no need for ``count``.

File: lib/system.nim (hunk 3 above)
Comment (timotheecour) [reply lib/system.nim]: that's the hack I'm using to detect named vs un-named tuple reliably:
```nim
  let a1 = (Field0:"foo",)
  echo a1.Field0 # works
  let a2 = ("foo",)
  # echo a2.Field0 # fails
```

another option would be to add a magic.

File: lib/system.nim (hunk 5 above)
Comment (timotheecour) [reply lib/system.nim]: not super sure what you're suggesting here; I wanna avoid duplicating this snippet:
```
    when compiles($value):
      when value isnot string and value isnot seq and compiles(value.isNil):
        if value.isNil: result.add "nil"
        else: result.addQuoted(value)
      else:
        result.addQuoted(value)
      firstElement = false
    else:
      result.add("...")
```

I could use a template to wrap that and have 2 independent loops (1 for `isNamed` = true vs false), but the way I wrote it seems simpler

Comment (Araq) [reply lib/system.nim]: hmm ok.

File: compiler/types.nim (hunk 2 above)
Comment (Araq) [reply compiler/types.nim]: Why did it work before? The compiler code is quite clear, you cannot put a static into a tuple.

Comment (timotheecour) [reply compiler/types.nim]: > Why did it work before? 

because support for it had been added here: https://github.com/nim-lang/Nim/pull/9046, but that logic failed after current PR added ` `compiles(...)` inside `proc `$`*[T: tuple|object](x: T): string =`

I pushed a commit that reverts the change in compiler/types.nim and modifies ./tests/vm/tissues.nim to not rely on having a static in a tuple, marking as resolved.

File: lib/system.nim [hunk 7]
```diff
@@ -2648,19 +2648,40 @@ proc compiles*(x: untyped): bool {.magic: "Compiles", noSideEffect, compileTime.
   ##     echo "'+' for integers is available"
   discard
 
+proc isNamedTuple*(T: type): bool =
```
Comment (GULPF) [inline lib/system.nim]: Can't we avoid adding this to system.nim? It could be moved to `helpers.nim` and exported in `typetraits`, `sugar` or some other fitting module.

File: lib/system.nim (hunk 3 above)
Comment (krux02) [reply lib/system.nim]: I don't like when anything is based on the `compiles` attribute. For my personal preference you should add a magic, but I think @Araq has a different opinion about it.

File: lib/system.nim (hunk 7 above)
Comment (timotheecour) [reply lib/system.nim]: fixed

**Comment by timotheecour** [comment]
PTAL, moved isNamedTuple to helpers.nim, now included by system.nim and typetraits.nim

> Generally I like the changes. As I said, I don't like the hack with compiles, but apart from that I approve this PR.

if it becomes a CT performance bottleneck (which I doubt, but we can now measure, see https://github.com/nim-lang/Nim/issues/10132), we can always change implementation later to use a magic proc instead of `compiles`

**Comment by timotheecour** [comment]
> Generally I like the changes. As I said, I don't like the hack with compiles, but apart from that I approve this PR.

merging since all comments were addressed and it was approved; as said earlier, we can always introduce a `magic` instead of `compiles` if ever needed in a subsequent PR
